### DISCUSSION TITLE
Is it possible to automatically include all data files given in setup.py section package_data of dependencies?

### DISCUSSION THREAD

**Discussion by Eugene-1984** [pyinstaller/pyinstaller#7910] [Help] [answered]
Hello!

I am packing an app that depends on other packages, and these packages often use package_data section in their setup.py files.

Thus I have to manually add Analysis.datas entries in my spec file (or semi-manually using Tree class, but still for each directory manually).

If a dependency has package_data filed in its setup.py is it not logical to automatically include all these files in a bundle? Since if I depend on that package and that packages requires extra files in its package_data to work properly, then I also want these files.

Is there a way to make it all happen for all dependencies? Or at least list dependencies manually and then scan their package_data contents and put them all in a bundle also?

**Comment by rokm** [comment]
Not out of the box. 

PyInstaller intentionally does not collect packages' data and metada - whether the bundled data files are actually required by the package is highly dependent on the package. Sometimes they contain example or test dataset files, or data files are used only by specific modules that are not used by the frozen application. Therefore, if the data is required, it is up to you to ensure it is collected - either via [hooks](https://github.com/pyinstaller/pyinstaller-hooks-contrib/tree/master/src/_pyinstaller_hooks_contrib/hooks/stdhooks) or via .spec file.

> Thus I have to manually add Analysis.datas entries in my spec file (or semi-manually using Tree class, but still for each directory manually).

The only legitimate reason for editing `Analysis.datas` is if you want to *exclude* some files.

If you want to add data files, the preferred way is to use the [`collect_data_files` hook utility function](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_data_files), and pass the resulting list of tuples to `Analysis`'s constructor via `datas` argument.

The `collect_data_files` operates on *package* names (not dist names!); so you could create a loop over list of used packages and call the helper to construct the list of data files (similar to what you are already doing with `Tree`, but without hardcoded paths to packages or the environment prefix). You could also create a helper function that takes the list of dist names, reads their metadata, and constructs the list of `datas` tuples to pass to `Analysis`' constructor.

**Comment by bwoodsend** [comment, marked as answer]
If you're trying to dynamically collect everything from every package, that's usually a sign that you're overthinking it. Usually you never need more than a couple of `--collect-data=package1 --collect-data=package2` flags. Trying to blindly collect everything usually results in applications full of test suites, example data, use case specific files and other rubbish which you then have to comb through manually and selectively remove.

**Comment by Eugene-1984** [comment]
Thank you for the answers!

It may be of some help to modify lines 
> And the command line equivalent (see [What To Bundle, Where To Search](https://pyinstaller.org/en/stable/usage.html#what-to-bundle-where-to-search) for platform-specific details):
in https://pyinstaller.org/en/stable/spec-files.html#adding-data-files to emphasize that the "What To Bundle, Where To Search" article covers no only platform-specific details but also a wider range of automatic collection options.
